app/services/testing_ground/save_as.rb

```diff
@@ -18,14 +18,16 @@ def initialize(original, name, user = nil)
 
   def run
     duplicate = @original.dup
+    duplicate.name = @name
+    duplicate.user = @user
 
-    TestingGround.transaction do
-      duplicate.name = @name
-      duplicate.user = @user
+    if duplicate.valid?
```

Comment:
The whitespace on these lines is a bit wonky; there are a couple of blank lines you probably didn't mean to commit:

```ruby
if duplicate.valid?
  TestingGround.transaction do

    duplicate.save!

    duplicate_associations!(duplicate)
  end
end
```

It's not really necessary to run `#valid?` here since `save!` checks the object validity already (these three lines effectively run `duplicate#valid?` twice). It also possible that one of the duplicated associations raises an error due to being invalid, and `duplicate#valid?` won't catch that.

An alternative would be to skip the `valid?`, let the exception get raised, and catch it in the controller like so:

```ruby
def save_as
  @testing_ground = TestingGround::SaveAs.run(
    @testing_ground, testing_ground_params[:name], current_user
  )

  render json: { redirect: testing_ground_path(@testing_ground) }
catch ActiveRecord::RecordInvalid
  render json: { errors: @testing_ground.errors }, status: 422
end
```